docs/zkapps/o1js/bitwise-operations.mdx

```diff
@@ -0,0 +1,456 @@
+
+A bitwise NOT operation returns 1 in each bit position if the corresponding bit of the operand is 0 and returns 0 if the corresponding bit of the operand is 1.
+
+The NOT operation:
+
+- Fails if the length or the input value is larger than 254 bits.
+- Operates only over the number of bits specified by the `length` parameter.
+    - The `length` parameter defines how many bits to NOT.
+    - A larger `length` parameter adds more constraints.
+- Is implemented in two different ways:
+    - If the `checked` parameter is set to `false`, it is implemented as a subtraction of the input from the all-ones bitmask (all the bits in a binary sequence are set to 1). This approach is more efficient, but does not check the length of the input. If the `checked` parameter is not provided, the default implementation uses `false`.
+    - If the `checked` parameter is set to `true`, the [xor()](#xor) gadget is reused with a second argument to be an all-ones bitmask of the same length. This approach is less efficient, needing as many rows as an XOR requires for a single negation, but comes with the advantage of making sure the input is of a certain length.
+
+For implementation details, see [NOT](https://o1-labs.github.io/proof-systems/specs/kimchi.html?highlight=gates#not) in the Mina book.
+
+Example NOT-ing 4 bits with the unchecked version:
+
+```ts
+let a = Field(0b0101);
+let b = Gadgets.not(a,4,false);
+
+b.assertEquals(0b1010);
+```
+
+Example NOT-ing 4 bits with the checked version using `xor()`:
+
+```ts
+let a = Field(0b0101);
+let b = Gadgets.not(a,4,true);
+
+b.assertEquals(0b1010);
+```
+
+## xor()
+
+```ts
+xor(a: Field, b: Field, length: number) => Field
+```
+
+The `xor()` gadget is equivalent to the [Bitwise XOR (^)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Bitwise_XOR) operator in JavaScript. It is a provable method to support bitwise XOR operations for native Field elements. 
+
+The XOR operation:
+
+- Compares two bits and returns 1 if two bits differ and returns 0 if two bits are equal.
+- Recursively builds a chain of XOR gates. 
+- Verifies that each XOR gate can verify at most 16 bits.
+- Adds another XOR gate to the chain if your input elements exceed 16 bits.
+- Rounds `length` to the nearest multiple of 16, `paddedLength = ceil(length / 16) * 16`.
+- Constrains input values to fit into `paddedLength` bits. 
+- Requires that both Field elements fit into `2^paddedLength`.
+    For example, `length = 2` (`paddedLength = 16`) fails for any input that is larger than `2**16`.
+- Guarantees the output has at most `paddedLength` bits.
+
+The `length` parameter defines how many bits to compare. When choosing the value of the `length` parameter, be sure to carefully weigh the trade-off between an increased number of constraints and security:
```

Comment:
we have a variety of ways we talk about the length parameter. Would it make sense to expand the length parameter discussion for the other gadgets to be more like this one?

Reply:
We'd need to confirm whether all length parameters result in additional constraints with larger values. Will defer to eng. review on this one.